qiskit/circuit/controlflow/if_else.py

```diff
@@ -154,6 +136,358 @@ def blocks(self):
 
     def c_if(self, classical, val):
         raise NotImplementedError(
-            "WhileLoopOp cannot be classically controlled through Instruction.c_if. "
-            "Please use an IfElseOp instead."
+            "IfElseOp cannot be classically controlled through Instruction.c_if. "
+            "Please nest it in an IfElseOp instead."
         )
+
+
+class IfElsePlaceholder(InstructionPlaceholder):
+    """A placeholder instruction to use in control-flow context managers, when calculating the
+    number of resources this instruction should block is deferred until the construction of the
+    outer loop.
+
+    This generally should not be instantiated manually; only :obj:`.IfContext` and
+    :obj:`.ElseContext` should do it when they need to defer creation of the concrete instruction.
+    """
+
+    def __init__(
+        self,
+        condition: Tuple[Union[ClassicalRegister, Clbit], int],
+        true_block: ControlFlowBuilderBlock,
+        false_block: Optional[ControlFlowBuilderBlock] = None,
+        *,
+        label: Optional[str] = None,
+    ):
+        """
+        Args:
+            condition: the condition to execute the true block on.  This has the same semantics as
+                the ``condition`` argument to :obj:`.IfElseOp`.
+            true_block: the unbuilt scope block that will become the "true" branch at creation time.
+            false_block: if given, the unbuilt scope block that will become the "false" branch at
+                creation time.
+            label: the label to give the operator when it is created.
+        """
+        # These are protected names because we're not trying to clash with parent attributes.
+        self.__true_block = true_block
+        self.__false_block: Optional[ControlFlowBuilderBlock] = false_block
+        self.__resources = self._placeholder_resources()
+        qubits, clbits = self.__resources
+        super().__init__("if_else", len(qubits), len(clbits), [], label=label)
+        # Set the condition after super().__init__() has initialised it to None.
+        self.condition = validate_condition(condition)
+
+    def with_false_block(self, false_block: ControlFlowBuilderBlock) -> "IfElsePlaceholder":
+        """Return a new placeholder instruction, with the false block set to the given value,
+        updating the bits used by both it and the true body, if necessary.
+
+        It is an error to try and set the false block on a placeholder that already has one.
+
+        Args:
+            false_block: The (unbuilt) instruction scope to set the false body to.
```

Comment:
Yeah, the type hint should indicate this sort of thing.  I didn't worry too much about _super_ defensive error checking on things like types - even something being technically the correct type in Python isn't sufficient to guarantee that'll work as expected, because you can `del` and override arbitrary methods on single instances at runtime without changing the type.